src/storage/src/source/mysql/snapshot.rs

```diff
@@ -0,0 +1,281 @@
+            let mut tx_opts = TxOpts::default();
+            tx_opts
+                .with_isolation_level(IsolationLevel::RepeatableRead)
+                .with_consistent_snapshot(true)
+                .with_readonly(true);
+            conn.start_transaction(tx_opts).await?;
+            conn.query_drop("set @@session.time_zone = '+00:00'")
+                .await?;
+
+            // Read the schemas of the tables we are snapshotting
+            // TODO: verify the schema matches the expected schema
+            let _ = schema_info(&mut conn, &SchemaRequest::Tables(reader_snapshot_table_info.keys().map(|f| (f.0[0].as_str(), f.0[1].as_str())).collect())).await?;
+
+            // Send a signal to the leader that we have started our transaction
+            transaction_start_handle.give(&transaction_start_cap_set[0], ()).await;
+            *transaction_start_cap_set = CapabilitySet::new();
+
+            trace!(%id, "timely-{worker_id} started transaction");
+
+            if is_snapshot_leader {
+                // Wait for all workers to start their transactions so we can release the table locks
+                let mut received = 0;
+                while received < config.worker_count {
+                    match transaction_start_input.next().await {
+                        Some(AsyncEvent::Data(_, _)) => received += 1,
+                        Some(AsyncEvent::Progress(_)) => (),
+                        None => panic!("transaction_start_input closed unexpectedly"),
+                    }
+                }
+
+                trace!(%id, "timely-{worker_id} releasing table locks");
+                if let Some(mut lock_conn) = lock_conn.take() {
+                    lock_conn.query_drop("UNLOCK TABLES").await?;
+                    lock_conn.disconnect().await?;
+                } else {
+                    panic!("lock_conn should have been created for the snapshot leader");
+                }
+                drop(lock_conn);
+            }
+
+            // We have established a snapshot GTID set so we can broadcast the rewind requests
+            for table in reader_snapshot_table_info.keys() {
+                trace!(%id, "timely-{worker_id} producing rewind request for {table}");
+                let req = RewindRequest { table: table.clone(), snapshot_gtid_set: snapshot_gtid_set.clone() };
+                rewinds_handle.give(&rewind_cap_set[0], req).await;
+            }
+            *rewind_cap_set = CapabilitySet::new();
+
+            // Read the snapshot data from the tables
+            let mut final_row = Row::default();
+            for (table, (output_index, table_desc)) in reader_snapshot_table_info {
+                let query = format!("SELECT * FROM {}", table.to_ast_string());
+                trace!(%id, "timely-{worker_id} reading snapshot from table: {table}\n{table_desc:?}");
+                let mut results = conn.exec_stream(query, ()).await?;
```

Comment:
Yep if we want to do the count concurrently we'll have to create a second client and open the transaction the same way (before the main worker releases the table locks). There is already one client/connection per-worker + the lock-connection so this would almost double that. Unless we just used one extra client for getting the counts for all tables?

Reply:
Ideally we would do this soon (before GA) so that we bake in the performance cost of doing counting, certainly not in this pr tho! (with postgres we have to tack it on after the fact, which is why I am doing that change before all the other metric stuff is setup)

For now in pg, and I imagine in mysql, I am doing an additional connection-per-worker. We didn't see issues when we went from 1 connection total to 1 connection per worker in postgres, and I imagine it won't be a problem in mysql as well. We can definitely change it in the future, but the general latency cost of counting in postgres AND mysql makes me want to start with as much concurrency as possible!

We can definitely make all this LD-configurable!